### ISSUE TITLE
Integration of object detection model

### ISSUE THREAD

**Issue by andreanne-lemay** [neuropoly/ivado-medical-imaging#218]
Currently, SCT has different machine learning and deep learning models. These models don't currently have a framework to easily retrain them on custom data (or the models are in Tensorflow). Ivado-medical-imaging offers a framework to train models, so integrating an object detection model could solve this issue. This way, we could have a flexible model to detect custom ROI (e.i. the spinal cord). The output of object detection models is a bounding box, so a second model could then be applied to the ROI contained in the bounding box.

Popular object detection models:
- YOLOv3 
- Faster RCNN
- SSD 

These models are 2D object detection models which can be an issue. Recombining the 2D boxes in one 3D bounding box is not always trivial. A model directly generating a 3D bounding box could be more fitted for medical images.

**Comment by jcohenadad** [comment]
awesome suggestion! to list some concrete examples of how this could be used:
- detection of ponto-medullary junction (currently `sct_detect_pmj` @charleygros )
- detection of C2-C3 disc (@lrouhier)
- detection of spinal cord on sagittal data, to be followed by a cropping and size-adjusted CNN for tumor detection (@andreanne-lemay)

I'm sure we can think of other applications

**Comment by charleygros** [comment]
> awesome suggestion! to list some concrete examples of how this could be used:
> 
>     * detection of ponto-medullary junction (currently `sct_detect_pmj` @charleygros )
> 
>     * detection of C2-C3 disc (@lrouhier)
> 
>     * detection of spinal cord on sagittal data, to be followed by a cropping and size-adjusted CNN for tumor detection (@andreanne-lemay)
> 
> 
> I'm sure we can think of other applications

Could add [Figure 3](https://arxiv.org/pdf/1805.06349.pdf).

**Comment by charleygros** [comment]
@jcohenadad --> would it be relevant to integrate [OptiC training pipeline](https://github.com/neuropoly/spinalcordtoolbox/blob/master/dev/detect_c2c3/train.py)?
The framework does exit.

**Comment by charleygros** [comment]
A crucial point would be to allow **end-to-end training** if this module is used for initialising a segmentation tool.

**Comment by charleygros** [comment]
We could start implementing a simple DenseNet for Detection baseline?
It would also imply to create a module of patch generation that will be plug to the current loader :)

**Comment by andreanne-lemay** [comment]
> A crucial point would be to allow **end-to-end training** if this module is used for initialising a segmentation tool.

A end-to-end training would be more convenient, but this can be challenging. If the segmentation model depends on the input of the first model (general detection), I believe we would still need to train the first part of the model first. Also, if they have different types of outputs (e.i. segmentation mask and box coordinates), we have to deal with the 2 outputs when training (2 separate backprops or combination of the 2 outputs as one output and a single backprop?).

**Comment by andreanne-lemay** [comment]
> We could start implementing a simple DenseNet for Detection baseline?
> It would also imply to create a module of patch generation that will be plug to the current loader :)

What type of output did you have in mind for the DenseNet? Did you have in mind to use the DenseNet to output a semantic segmentation and then generate a bounding box around this mask? Advantage I see to this approach: it limits the changes we'd have to do in the code, the output is a mask, just like UNet (no need to modify the loader). Disadvantage: less direct.
Personally, I had in mind to start with a 2D object detection model for detection baseline. If the results are encouraging, I'd then try to modify the model to generate a 3D bounding boxes. This approach needs more code addition, but can directly generate a bounding box.
There might also be other approaches I didn't consider. Any thoughts?

**Comment by lrouhier** [comment]
> A end-to-end training would be more convenient, but this can be challenging. If the segmentation model depends on the input of the first model (general detection), I believe we would still need to train the first part of the model first. Also, if they have different types of outputs (e.i. segmentation mask and box coordinates), we have to deal with the 2 outputs when training (2 separate backprops or combination of the 2 outputs as one output and a single backprop?).

The advantage of separating that in two different models is that if you have multiple applications that will use spinal cord localization (for example) you will already have part of your pipeline ready.

**Comment by andreanne-lemay** [comment]
> > A end-to-end training would be more convenient, but this can be challenging. If the segmentation model depends on the input of the first model (general detection), I believe we would still need to train the first part of the model first. Also, if they have different types of outputs (e.i. segmentation mask and box coordinates), we have to deal with the 2 outputs when training (2 separate backprops or combination of the 2 outputs as one output and a single backprop?).
> 
> The advantage of separating that in two different models is that if you have multiple applications that will use spinal cord localization (for example) you will already have part of your pipeline ready.

Yes I agree! Keeping 2 models, but automating the pipeline in a single training would be a good in between solution.

**Comment by andreanne-lemay** [comment]
Interesting repo for 3D detection: https://github.com/MIC-DKFZ/medicaldetectiontoolkit

**Comment by charleygros** [comment]
> End-to-end training could also mean having to 2 models trained one after the other, but one training would automatically be following the other. This approach might be more realistic. What did you have in mind :)?

The basic aspect that I had in mind was: to use the output of the detection module (previously trained) during the training of the second module (e.g. segmentation) --> the segmentation module is therefore trained on "realistic" outputs of the detection module (vs. perfectly cantered patches that could be extracted from the ground-truth).

**Comment by charleygros** [comment]
> What type of output did you have in mind for the DenseNet? Did you have in mind to use the DenseNet to output a semantic segmentation and then generate a bounding box around this mask?

I liked the idea of having an uncertainty that could be propagated to the second module (e.g. segmentation). But I imagine it could also be achieved with boundary boxes.

> Disadvantage: less direct.

Not sure to see what you mean here :) Could you please help me?

> Personally, I had in mind to start with a 2D object detection model for detection baseline. If the results are encouraging, I'd then try to modify the model to generate a 3D bounding boxes.

Yep for sure, sounds great :) ! Something to have in mind would be the size of the bounding boxes --> how to handle different input size for the second module? We will need to bridge this aspect!

**Comment by andreanne-lemay** [comment]
> > What type of output did you have in mind for the DenseNet? Did you have in mind to use the DenseNet to output a semantic segmentation and then generate a bounding box around this mask?
> 
> I liked the idea of having an uncertainty that could be propagated to the second module (e.g. segmentation). But I imagine it could also be achieved with boundary boxes.
> 
Yes, there is a confidence associated with the bounding boxes generated :) 
> > Disadvantage: less direct.
> 
> Not sure to see what you mean here :) Could you please help me?
> 

What I mean is that we require a segmentation as input to output a box + we don't generate directly the bounding boxes. I am not sure which approach is better for our needs. 

> > Personally, I had in mind to start with a 2D object detection model for detection baseline. If the results are encouraging, I'd then try to modify the model to generate a 3D bounding boxes.
> 
> Yep for sure, sounds great :) ! Something to have in mind would be the size of the bounding boxes --> how to handle different input size for the second module? We will need to bridge this aspect!

Yes definitely, but we can train with different size images (I didn't think so at first), so it might not be a problem!
